tests/bluetooth/bsim_bt/bsim_test_l2cap/src/main_l2cap_ecred.c

```diff
@@ -0,0 +1,618 @@
+		k_work_init(&chan->work, send_sdu_chan_worker);
+		chan->le.chan.ops = &l2cap_ops;
+		chan->le.rx.mtu = DATA_MTU;
+		chan->le.rx.mps = DATA_MPS;
+
+		return chan;
+	}
+
+	return NULL;
+}
+
+static void connect_num_channels(uint8_t num_l2cap_channels)
+{
+	struct channel *chan = NULL;
+	struct bt_l2cap_chan *allocated_channels[ECRED_CHAN_MAX] = { NULL };
+	int err;
+
+	for (int i = 0; i < num_l2cap_channels; i++) {
+		chan = get_free_channel();
+
+		if (!chan) {
+			FAIL("failed, chan not free");
+
+			return;
+		}
+
+		allocated_channels[i] = &chan->le.chan;
+	}
+
+	err = bt_l2cap_ecred_chan_connect(default_conn, allocated_channels, servers[0].psm);
+	if (err) {
+		FAIL("can't connect ecred %d ", err);
+	}
+}
+
+static void disconnect_all_channels(void)
+{
+	for (int i = 0; i < ARRAY_SIZE(channels); i++) {
+		if (channels[i].in_use) {
+			int err;
+
+			LOG_DBG("Disconnecting channel: %d)", channels[i].chan_id);
+			err = bt_l2cap_chan_disconnect(&channels[i].le.chan);
+
+			if (err) {
+				LOG_DBG("can't disconnnect channel (err: %d)", err);
+			}
+
+			channels[i].in_use = false;
+		}
+	}
+}
+
+static int accept(struct bt_conn *conn, struct bt_l2cap_chan **l2cap_chan)
+{
+	struct channel *chan;
+
+	chan = get_free_channel();
+	if (!chan) {
+
+		return -ENOMEM;
+	}
+
+	*l2cap_chan = &chan->le.chan;
+
+	return 0;
+}
+
+static struct bt_l2cap_server *get_free_server(void)
+{
+	for (int i = 0; i < SERVERS; i++) {
+		if (servers[i].psm) {
+
+			continue;
+		}
+
+		return &servers[i];
+	}
+
+	return NULL;
+}
+
+static void register_l2cap_server(void)
+{
+	struct bt_l2cap_server *server;
+
+	server = get_free_server();
+	if (!server) {
+		FAIL("Failed to get free server");
+		return;
+	}
+
+	server->accept = accept;
+	server->psm = 0;
+
+	if (bt_l2cap_server_register(server) < 0) {
+		FAIL("Failed to get free server");
+
+		return;
+	}
+
+	LOG_DBG("L2CAP server registered, PSM:0x%X", server->psm);
+}
+
+static void connected(struct bt_conn *conn, uint8_t conn_err)
+{
+	char addr[BT_ADDR_LE_STR_LEN];
+
+	bt_addr_le_to_str(bt_conn_get_dst(conn), addr, sizeof(addr));
+
+	if (conn_err) {
+		FAIL("Failed to connect to %s (%u)", addr, conn_err);
+		bt_conn_unref(default_conn);
+		default_conn = NULL;
+
+		return;
+	}
+
+	default_conn = bt_conn_ref(conn);
+	LOG_DBG("%s", addr);
+
+	SET_FLAG(is_connected);
+}
+
+static void disconnected(struct bt_conn *conn, uint8_t reason)
+{
+	char addr[BT_ADDR_LE_STR_LEN];
+
+	bt_addr_le_to_str(bt_conn_get_dst(conn), addr, sizeof(addr));
+
+	LOG_DBG("%s (reason 0x%02x)", addr, reason);
+
+	if (default_conn != conn) {
+		FAIL("Conn mismatch disconnect %s %s)", default_conn, conn);
+
+		return;
+	}
+
+	bt_conn_unref(default_conn);
+	default_conn = NULL;
+	UNSET_FLAG(is_connected);
+}
+
+BT_CONN_CB_DEFINE(conn_callbacks) = {
+	.connected = connected,
+	.disconnected = disconnected,
+};
+
+static void send_sdu(int iteration, int chan_idx, int bytes);
+
+void send_sdu_chan_worker(struct k_work *item)
```

Comment:
If you move `send_sdu` to before `send_sdu_chan_worker` you can avoid the prototype. In any case, please put function prototypes to top of file.